{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2130231f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scripts import helpers, mlogo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3453a20b",
   "metadata": {},
   "outputs": [],
   "source": [
    "helpers.read_json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7b77942",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Modules import\n",
    "import os \n",
    "import pandas as pd\n",
    "from scripts import helpers\n",
    "\n",
    "# Class that generate motif-logo and perform data analysis on DNA sequences\n",
    "class mlogo():\n",
    "    \"\"\"\n",
    "    Class that generate motif-logo and perform data analysis on DNA sequences.\n",
    "    The data can be grouped by the \"group_by\" argument (column) and generate \n",
    "    seperate plot for each subject \n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self,\n",
    "                 seq_file,\n",
    "                 group_by:str = None,\n",
    "                 divide_subject:bool = True,\n",
    "                 subject_col:str = \"subject_id\",\n",
    "                 sequence_col:str = \"sequence\",\n",
    "                 germline_col:str = \"germline\"):\n",
    "        \"\"\"\n",
    "        seq_file : str -> file path / df object of sequencing data.\n",
    "        group_by : str -> group the dataset by this column. will create number of subplots according\n",
    "                          to the number of unique values in the grouped by column.\n",
    "        divide_subject : bool -> Generate different plot for each subject under `subject_id` column.\n",
    "        subject_col : str -> name of the subject column, defualt is `subject_id`. \n",
    "        sequence_col : str -> name of the DNA sequence column, defualt it `sequence`.\n",
    "        germline_col : str -> name of the germline DNA sequence column, defualt it `germline`.\n",
    "        \"\"\"\n",
    "        \n",
    "        # Loading sequences dataset\n",
    "        try:\n",
    "            if isinstance(seq_file, str):        \n",
    "                self.seq_df = pd.read_csv(seq_file, index_col=0)\n",
    "                print(f\"> Dataset loaded (seq_file = '{seq_file}')\")\n",
    "\n",
    "            elif isinstance(seq_file, pd.DataFrame):\n",
    "                 self.seq_df = seq_file\n",
    "                 print(\"> Dataset loaded\")\n",
    "\n",
    "        except:\n",
    "                print(f\"> Invalid input, please make sure seq_file argument entred correctly.\\n  (Invalid: seq_file = '{seq_file}')\")\n",
    "\n",
    "        # Translating germline and sequence DNA into amino acid sequence\n",
    "        self.seq_df[\"sequence_aa\"] = self.seq_df[\"sequence\"].apply(helpers.nt_transalte_104) # Somatic sequence\n",
    "        self.seq_df[\"germline_aa\"] = self.seq_df[\"germline\"].apply(helpers.nt_transalte_104) # Germline sequence\n",
    "        \n",
    "\n",
    "        def motif_logo(self):\n",
    "             pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "996d1f76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Dataset loaded\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>ai</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>clone_id</th>\n",
       "      <th>functional</th>\n",
       "      <th>copy_number</th>\n",
       "      <th>cdr3_aa</th>\n",
       "      <th>sequence</th>\n",
       "      <th>germline</th>\n",
       "      <th>ab_target</th>\n",
       "      <th>time_point</th>\n",
       "      <th>label</th>\n",
       "      <th>sequence_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17877</th>\n",
       "      <td>M03592:154:000000000-JCY9V:1:2117:16836:3379</td>\n",
       "      <td>257367</td>\n",
       "      <td>8</td>\n",
       "      <td>subj_3</td>\n",
       "      <td>651989</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>CATDFNSNKAHW</td>\n",
       "      <td>.................................................</td>\n",
       "      <td>CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...</td>\n",
       "      <td>sn</td>\n",
       "      <td>4</td>\n",
       "      <td>sn.4.subj_3</td>\n",
       "      <td>[-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17880</th>\n",
       "      <td>M03592:154:000000000-JCY9V:1:2119:17338:11554</td>\n",
       "      <td>310211</td>\n",
       "      <td>10</td>\n",
       "      <td>subj_3</td>\n",
       "      <td>651991</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>CTQGAVAGKPEYFHNX</td>\n",
       "      <td>.................................................</td>\n",
       "      <td>CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...</td>\n",
       "      <td>sp</td>\n",
       "      <td>1</td>\n",
       "      <td>sp.1.subj_3</td>\n",
       "      <td>[-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17881</th>\n",
       "      <td>M03592:154:000000000-JCY9V:1:2119:16221:19845</td>\n",
       "      <td>310212</td>\n",
       "      <td>10</td>\n",
       "      <td>subj_3</td>\n",
       "      <td>651991</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>CTQGAVAGKPEYFHNX</td>\n",
       "      <td>.................................................</td>\n",
       "      <td>CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...</td>\n",
       "      <td>sp</td>\n",
       "      <td>1</td>\n",
       "      <td>sp.1.subj_3</td>\n",
       "      <td>[-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17882</th>\n",
       "      <td>M03592:154:000000000-JCY9V:1:2117:14781:11724</td>\n",
       "      <td>310217</td>\n",
       "      <td>10</td>\n",
       "      <td>subj_3</td>\n",
       "      <td>651991</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>CTQGAVAGKPEYFXNW</td>\n",
       "      <td>.................................................</td>\n",
       "      <td>CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...</td>\n",
       "      <td>sp</td>\n",
       "      <td>1</td>\n",
       "      <td>sp.1.subj_3</td>\n",
       "      <td>[-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17883</th>\n",
       "      <td>M03592:154:000000000-JCY9V:1:2114:22937:3453</td>\n",
       "      <td>310221</td>\n",
       "      <td>10</td>\n",
       "      <td>subj_3</td>\n",
       "      <td>651991</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>CTQGAVAGKPEXFXNW</td>\n",
       "      <td>.................................................</td>\n",
       "      <td>CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...</td>\n",
       "      <td>sp</td>\n",
       "      <td>1</td>\n",
       "      <td>sp.1.subj_3</td>\n",
       "      <td>[-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19031</th>\n",
       "      <td>M03592:154:000000000-JCY9V:1:2112:12328:4442</td>\n",
       "      <td>339103</td>\n",
       "      <td>12</td>\n",
       "      <td>subj_3</td>\n",
       "      <td>652294</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>CARAGXGSWSLIXXW</td>\n",
       "      <td>.................................................</td>\n",
       "      <td>CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...</td>\n",
       "      <td>sp</td>\n",
       "      <td>2</td>\n",
       "      <td>sp.2.subj_3</td>\n",
       "      <td>[-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19032</th>\n",
       "      <td>M03592:154:000000000-JCY9V:1:2110:21993:10357</td>\n",
       "      <td>339106</td>\n",
       "      <td>12</td>\n",
       "      <td>subj_3</td>\n",
       "      <td>652294</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>CXRADFGSWSLIDNW</td>\n",
       "      <td>.................................................</td>\n",
       "      <td>CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...</td>\n",
       "      <td>sp</td>\n",
       "      <td>2</td>\n",
       "      <td>sp.2.subj_3</td>\n",
       "      <td>[-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19033</th>\n",
       "      <td>M03592:154:000000000-JCY9V:1:2110:15095:22503</td>\n",
       "      <td>339110</td>\n",
       "      <td>12</td>\n",
       "      <td>subj_3</td>\n",
       "      <td>652294</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>CARAXFGSXSLIDNW</td>\n",
       "      <td>.................................................</td>\n",
       "      <td>CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...</td>\n",
       "      <td>sp</td>\n",
       "      <td>2</td>\n",
       "      <td>sp.2.subj_3</td>\n",
       "      <td>[-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19034</th>\n",
       "      <td>M03592:154:000000000-JCY9V:1:2110:12209:13507</td>\n",
       "      <td>339111</td>\n",
       "      <td>12</td>\n",
       "      <td>subj_3</td>\n",
       "      <td>652294</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>CARADFGXRSLIDNW</td>\n",
       "      <td>.................................................</td>\n",
       "      <td>CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...</td>\n",
       "      <td>sp</td>\n",
       "      <td>2</td>\n",
       "      <td>sp.2.subj_3</td>\n",
       "      <td>[-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19035</th>\n",
       "      <td>M03592:154:000000000-JCY9V:1:2110:11570:10387</td>\n",
       "      <td>339112</td>\n",
       "      <td>12</td>\n",
       "      <td>subj_3</td>\n",
       "      <td>652294</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>CARADFGSWSLIDNW</td>\n",
       "      <td>.................................................</td>\n",
       "      <td>CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...</td>\n",
       "      <td>sp</td>\n",
       "      <td>2</td>\n",
       "      <td>sp.2.subj_3</td>\n",
       "      <td>[-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              seq_id      ai  sample_id  \\\n",
       "17877   M03592:154:000000000-JCY9V:1:2117:16836:3379  257367          8   \n",
       "17880  M03592:154:000000000-JCY9V:1:2119:17338:11554  310211         10   \n",
       "17881  M03592:154:000000000-JCY9V:1:2119:16221:19845  310212         10   \n",
       "17882  M03592:154:000000000-JCY9V:1:2117:14781:11724  310217         10   \n",
       "17883   M03592:154:000000000-JCY9V:1:2114:22937:3453  310221         10   \n",
       "...                                              ...     ...        ...   \n",
       "19031   M03592:154:000000000-JCY9V:1:2112:12328:4442  339103         12   \n",
       "19032  M03592:154:000000000-JCY9V:1:2110:21993:10357  339106         12   \n",
       "19033  M03592:154:000000000-JCY9V:1:2110:15095:22503  339110         12   \n",
       "19034  M03592:154:000000000-JCY9V:1:2110:12209:13507  339111         12   \n",
       "19035  M03592:154:000000000-JCY9V:1:2110:11570:10387  339112         12   \n",
       "\n",
       "      subject_id  clone_id  functional  copy_number           cdr3_aa  \\\n",
       "17877     subj_3    651989           1            4      CATDFNSNKAHW   \n",
       "17880     subj_3    651991           1            9  CTQGAVAGKPEYFHNX   \n",
       "17881     subj_3    651991           1            3  CTQGAVAGKPEYFHNX   \n",
       "17882     subj_3    651991           1            3  CTQGAVAGKPEYFXNW   \n",
       "17883     subj_3    651991           1            4  CTQGAVAGKPEXFXNW   \n",
       "...          ...       ...         ...          ...               ...   \n",
       "19031     subj_3    652294           1            2   CARAGXGSWSLIXXW   \n",
       "19032     subj_3    652294           1            3   CXRADFGSWSLIDNW   \n",
       "19033     subj_3    652294           1            2   CARAXFGSXSLIDNW   \n",
       "19034     subj_3    652294           1            3   CARADFGXRSLIDNW   \n",
       "19035     subj_3    652294           1            3   CARADFGSWSLIDNW   \n",
       "\n",
       "                                                sequence  \\\n",
       "17877  .................................................   \n",
       "17880  .................................................   \n",
       "17881  .................................................   \n",
       "17882  .................................................   \n",
       "17883  .................................................   \n",
       "...                                                  ...   \n",
       "19031  .................................................   \n",
       "19032  .................................................   \n",
       "19033  .................................................   \n",
       "19034  .................................................   \n",
       "19035  .................................................   \n",
       "\n",
       "                                                germline ab_target  \\\n",
       "17877  CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...        sn   \n",
       "17880  CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...        sp   \n",
       "17881  CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...        sp   \n",
       "17882  CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...        sp   \n",
       "17883  CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...        sp   \n",
       "...                                                  ...       ...   \n",
       "19031  CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...        sp   \n",
       "19032  CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...        sp   \n",
       "19033  CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...        sp   \n",
       "19034  CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...        sp   \n",
       "19035  CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...        sp   \n",
       "\n",
       "       time_point        label  \\\n",
       "17877           4  sn.4.subj_3   \n",
       "17880           1  sp.1.subj_3   \n",
       "17881           1  sp.1.subj_3   \n",
       "17882           1  sp.1.subj_3   \n",
       "17883           1  sp.1.subj_3   \n",
       "...           ...          ...   \n",
       "19031           2  sp.2.subj_3   \n",
       "19032           2  sp.2.subj_3   \n",
       "19033           2  sp.2.subj_3   \n",
       "19034           2  sp.2.subj_3   \n",
       "19035           2  sp.2.subj_3   \n",
       "\n",
       "                                             sequence_aa  \n",
       "17877  [-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...  \n",
       "17880  [-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...  \n",
       "17881  [-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...  \n",
       "17882  [-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...  \n",
       "17883  [-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...  \n",
       "...                                                  ...  \n",
       "19031  [-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...  \n",
       "19032  [-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...  \n",
       "19033  [-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...  \n",
       "19034  [-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...  \n",
       "19035  [-, -, -, -, -, -, -, -, -, -, -, -, -, -, -, ...  \n",
       "\n",
       "[1000 rows x 14 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df = pd.read_csv(\"input\\\\cl_seqs_motif.csv\", index_col=0).head(1000)\n",
    "test = mlogo(seq_file=test_df)\n",
    "test.seq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "58f9cca5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MREIDN'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from scripts.helpers import nt_transalte_104\n",
    "seq = \"ATGCGTGAGATCGATAATAAT\"\n",
    "\"\".join(nt_transalte_104(seq))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "e70c7534",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Description of example matrix \"nn_saliency_matrix\":\n",
      "# \n",
      "# Saliency matrix illustrated in Figure 1F.\n",
      "# Data are from Figure 1D of Janganathan et al. (2019),\n",
      "# and were kindly provided by Kyle Farh and Kishore Jaganathan.\n",
      "# \n",
      "# References:\n",
      "# \n",
      "# Jaganathan K et al. (2019) Predicting Splicing from Primary Sequence with\n",
      "# Deep Learning. Cell. 176(3):535â€“548.e24.\n",
      "# \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pos</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.001725</td>\n",
       "      <td>-0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.033557</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.030026</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012748</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.019015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.010646</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>129 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       A         C         G         T\n",
       "pos                                   \n",
       "0   -0.0 -0.000000 -0.001725 -0.000000\n",
       "1    0.0  0.000000  0.033557  0.000000\n",
       "2    0.0  0.000000  0.030026  0.000000\n",
       "3    0.0  0.000000  0.012748  0.000000\n",
       "4    0.0  0.000000  0.000337  0.000000\n",
       "..   ...       ...       ...       ...\n",
       "124  0.0  0.000000  0.000000  0.021006\n",
       "125  0.0  0.000000  0.000000  0.019015\n",
       "126  0.0  0.000000  0.000000  0.010700\n",
       "127  0.0  0.000000  0.000000  0.010441\n",
       "128 -0.0 -0.010646 -0.000000 -0.000000\n",
       "\n",
       "[129 rows x 4 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import logomaker\n",
    "logomaker.get_example_matrix(\"nn_saliency_matrix\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "771e9851",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scripts.helpers import read_json\n",
    "\n",
    "def create_folders():\n",
    "    config_dict = read_json()\n",
    "    main_folders = [config_dict[\"input_folder\"], config_dict[\"output_folder\"]]\n",
    "    subf_output = [config_dict[\"output_folder\"] + i for i in [\"//motif_figure\", \"//motif_data`\"]]\n",
    "\n",
    "    for folder in main_folders + subf_output:\n",
    "        if os.path.exists(folder) is False:\n",
    "            os.mkdir(folder)\n",
    "            print(f\"> folder `{folder}` was created.\")\n",
    "\n",
    "\n",
    "create_folders()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
